test/puppetlabs/http/client/async_plaintext_test.clj

```diff
@@ -352,4 +356,95 @@
+                                (promise))]
+      (testwebserver/with-test-webserver-and-config
+       (build-content-handler data initial-bytes-read? nil) port {:shutdown-timeout-seconds 1}
+       (with-open [client (async/create-client {:connect-timeout-milliseconds 1000 :socket-timeout-milliseconds 2000})]
+         (let [response @(common/get
+                          client
+                          (str "http://localhost:" port "/hello")
+                          opts)
+               {:keys [status body]} response]
+           (is (= 200 status))
+           (let [instream body
+                 buf (make-array Byte/TYPE 4)
+                 _ (.read instream buf)]
+             ;; Make sure we can read a few chars off of the stream
+             (is (= "xxxx" (String. buf "UTF-8")))
+             ;; Indicate we read some chars
+             (if initial-bytes-read? (deliver initial-bytes-read? true))
+             ;; Read the rest and validate the content
+             (let [final-string (str "xxxx" (slurp instream))]
+               (is (= (str data "xxxx") final-string))))))))))
+
+(defn- clojure-streaming-failure
+  [data-size opts]
+  (testing (str "clojure streaming failure: " data-size " bytes with opts " opts)
+    (let [data (apply str (repeat (- data-size 4) "x"))]
+      (try
+        (testwebserver/with-test-webserver-and-config
+         (build-content-handler data nil true) port {:shutdown-timeout-seconds 1}
+         (with-open [client (async/create-client {:connect-timeout-milliseconds 1000 :socket-timeout-milliseconds 2000})]
+           (let [response @(common/get
+                            client
+                            (str "http://localhost:" port "/hello")
+                            opts)
+                 {:keys [body error]} response]
+             (if (= :unbuffered-stream (:as opts))
+               (do
+                 (if error
+                   ;; If there's an error this should behave the same as :stream
+                   (is (instance? SocketTimeoutException error))
+                   ;; If there is no error, let's consume the body to get the exception
+                   (is (thrown? SocketTimeoutException (slurp body)))))
+               (do
+                 (is error)
+                 (is (instance? SocketTimeoutException error)))))))
+        (catch TimeoutException e
+          ;; Expected whenever a server-side failure is generated
+          )))))
+
+(deftest clojure-streaming
+  (testing "clojure streaming is consistent with different payload sizes and opts"
+    (testlogging/with-test-logging
+     (doseq [data-size [32 (* 64 1024) (* 1024 1024)]
+             decompress-body? [false true]
+             as [:unbuffered-stream :stream]]
+       (clojure-streaming-success data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-failure data-size {:as as :decompress-body decompress-body?})))))
```

Comment:
I think it was unclear that I threw an Exception on the server side in the original commit. That exception is never actually propagated to the client, it just the server to stop responding and the Socket Timeout kicks in. I changed the code to just "wait forever" to make it more explicit that we're trying to invoke a SocketTimeoutException.

I also added a new test for ConnectException.

All we can is catch the exception from the underlying Socket on the client side. Ideally I'd like to implement some kind of read error where we can a few bytes back from the Socket then a sudden IOException but I can't find a way to force this without resorting to mocking.